cpp/src/decisiontree/batched-levelalgo/builder_base.cuh

```diff
@@ -451,11 +452,13 @@ struct RegTraits {
   static void computeSplit(Builder<RegTraits<DataT, IdxT>>& b, IdxT col,
                            IdxT batchSize, CRITERION splitType,
                            cudaStream_t s) {
-    auto n_col_blks = b.n_blks_for_cols;
+    auto n_col_blks = std::min(b.n_blks_for_cols, b.input.nSampledCols - col);
+
     dim3 grid(b.n_blks_for_rows, n_col_blks, batchSize);
     auto nbins = b.params.n_bins;
     size_t smemSize = 7 * nbins * sizeof(DataT) + nbins * sizeof(int);
     smemSize += sizeof(int);
+    smemSize += 7 * sizeof(DataT*);  // Room for alignment in worst case
```

Comment:
similar comment as mentioned above for classification.

Reply:
I wonder if there is any shortcut that would help here... right now this looks like we'll have to basically duplicate much of the allocation logic from kernels.cuh here and keep the files in sync forever with no tests in place to ensure we haven't messed up. Are we at risk of running out of shared memory here?

Reply:
I cleaned this up a bit in my latest push but didn't go all the way to the precise solution due to concerns above... let's discuss further if you think the memory overhead is an issue.